### PR TITLE
Freebsd-target-support

### PR DESCRIPTION
**Pull request by InfRandomness**
Implement freebsd as a target for miri

### CODE REVIEW

File: src/shims/unix/freebsd/mod.rs [hunk 1]
```diff
@@ -0,0 +1,2 @@
+pub mod foreign_items;
+pub mod dlsym;
```
Comment (bjorn3) [inline src/shims/unix/freebsd/mod.rs:2]: nit: missing trailing newline

File: src/shims/unix/macos/dlsym.rs [hunk 2]
```diff
@@ -3,21 +3,18 @@ use rustc_middle::mir;
 use log::trace;
 
 use crate::*;
-use helpers::check_arg_count;
 
 #[derive(Debug, Copy, Clone)]
 #[allow(non_camel_case_types)]
 pub enum Dlsym {
-    getentropy,
```
Comment (bjorn3) [inline src/shims/unix/macos/dlsym.rs:11]: Can you please revert the macos changes?

File: tests/pass/libc.rs [hunk 3]
```diff
@@ -5,12 +5,12 @@
 
 extern crate libc;
 
-#[cfg(target_os = "linux")]
+#[cfg(target_os = "linux, freebsd")]
```
Comment (bjorn3) [inline tests/pass/libc.rs:8]: The correct syntax is `#[cfg(any(target_os = "linux", target_os = "freebsd")]`.

File: diff [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+-    #[cfg(target_os = "linux")]
++    #[cfg(target_os = "linux,freebsd")]
+     test_prctl_thread_name();
+ 
+     test_thread_local_errno();
+ 
+-    #[cfg(target_os = "linux")]
++    #[cfg(target_os = "linux,freebsd")]
+     test_clocks();
+ }
```
[the remark this answers is not in the document]
Comment (InfRandomness) [reply diff:104]: oops

File: src/shims/env.rs [hunk 5]
```diff
@@ -55,7 +55,7 @@ impl<'tcx> EnvVars<'tcx> {
                 };
                 if forward {
                     let var_ptr = match target_os {
-                        "linux" | "macos" =>
+                        "linux" | "macos" | "freebsd" =>
```
Comment (RalfJung) [inline src/shims/env.rs:58]: As mentioned in the Android PR, this check should be moved into a separate function. Please make a separate PR that does only that.

Comment (InfRandomness) [reply src/shims/env.rs:58]: well I noticed that the android PR had the change already, so I didn't just want to yoink the diff from the android PR
but if you insist I can take their changes and make it into its own PR

**Comment by RalfJung** [comment]
Both PRs need that function - so landing it first is the best way to avoid conflicts later.

**Comment by RalfJung** [comment]
Thanks, looks like our target support can grow quite a bit in the near future. :)

I think for this to work we need to make sure that these targets can actually remain supported as the Rust standard library evolves. So if/when we add freebsd to CI and the supported target in the README, then we should also list there who is actually going to maintain that target -- so that we have someone to ping when there is trouble with that target.  We shouldn't have support for targets that nobody is going to maintain.  I don't want to be on the hook for maintaining them all. :)

So @InfRandomness do you want to be listed as maintaining freebsd support in Miri?

**Comment by InfRandomness** [comment]
> Thanks for the PR! Looks like our target support can grow quite a bit in the near future. :)
> 
> I think for this to work we need to make sure that these targets can actually remain supported as the Rust standard library evolves. So if/when we add freebsd to CI and the supported target in the README, then we should also list there who is actually going to maintain that target -- so that we have someone to ping when there is trouble with that target (the Rust standard library will inevitably change and that might require adjustments in Miri so that the target remains supported). We shouldn't have support for targets that nobody is going to maintain. And I don't want to be on the hook for maintaining them all. :)
> 
> So @InfRandomness do you want to be listed as maintaining freebsd support in Miri?

Sure !

**Comment by InfRandomness** [comment]
tests/pass/hello.rs seems to pass successfully but the other tests don't seem like they do

**Comment by dtolnay** [comment]
Thank you InfRandomness, this PR was already super helpful to me in https://github.com/rust-lang/libc/pull/2813#pullrequestreview-1012967919 for validating a FreeBSD codepath in the libc crate.

File: src/shims/unix/freebsd/foreign_items.rs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        _ret: mir::BasicBlock,
+    ) -> InterpResult<'tcx, EmulateByNameResult<'mir, 'tcx>> {
+        let this = self.eval_context_mut();
+        match link_name.as_str() {
+            // Querying system information
+            "pthread_attr_getstack" => {
+                // We don't support "pthread_attr_setstack", so we just pretend all stacks have the same values here.
+                let [attr_place, addr_place, size_place] =
+                    this.check_shim(abi, Abi::C { unwind: false }, link_name, args)?;
+                this.deref_operand(attr_place)?;
```
Comment (RalfJung) [inline src/shims/unix/freebsd/foreign_items.rs:26]:
```suggestion
                let _attr_place = this.deref_operand(attr_place)?;
```

Comment (RalfJung) [reply src/shims/unix/freebsd/foreign_items.rs:26]: Ah, this seems to actually be a standard POSIX thing? So probably best to move it to `shims/unix/foreign_items.rs` so Linux and FreeBSD can share it.

File: tests/pass/libc.rs [hunk 7]
```diff
@@ -290,26 +290,33 @@ fn test_clocks() {
     assert_eq!(is_error, 0);
 }
 
+#[cfg(target_os = "linux,freebsd,macos")]
+fn test_getpid() {
+    unsafe {
+        assert_eq!(libc::getpid(), std::process::id());
+    }
+}
```
Comment (RalfJung) [inline tests/pass/libc.rs:298]: This function is not even called; seems to be some rebase mistake?

**Comment by RalfJung** [comment]
This looks good overall! Please pull before pushing again; I have rebased your branch and adjusted CI to only run some minimal tests for FreeBSD. That should make CI pass, hopefully. :)

File: src/shims/unix/foreign_items.rs [hunk 8]
```diff
@@ -461,6 +461,28 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
                 this.write_null(dest)?;
             }
 
+            // Querying system information
+            "pthread_attr_getstack" => {
+                // We don't support "pthread_attr_setstack", so we just pretend all stacks have the same values here.
+                let [attr_place, addr_place, size_place] =
+                    this.check_shim(abi, Abi::C { unwind: false }, link_name, args)?;
+                this.deref_operand(attr_place)?;
```
Comment (RalfJung) [inline src/shims/unix/foreign_items.rs:469]:
```suggestion
                // We don't support "pthread_attr_setstack", so we just pretend all stacks have the same values here. Hence we can mostly ignore the input `attr_place`.
                let [attr_place, addr_place, size_place] =
                    this.check_shim(abi, Abi::C { unwind: false }, link_name, args)?;
                let _attr_place = this.deref_operand(attr_place)?;
```
